pkg/controllers/manila/manila.go

```diff
@@ -0,0 +1,221 @@
+//
+// Note that the CSI driver(s) are not un-installed when Manila becomes
+// missing or it stops providing shares of given type - Manila bight be
+// under (short?) maintenance / reconfiguration.
+// Similarly, StorageClasses are not deleted when a share type disappears
+// from Manila.
+type Controller struct {
+	operatorClient     v1helpers.OperatorClient
+	kubeClient         kubernetes.Interface
+	secretLister       corelisters.SecretLister
+	openStackClient    *openStackClient
+	storageClassLister storagelisters.StorageClassLister
+	// Controllers to start when Manila is detected
+	csiControllers     []Runnable
+	controllersRunning bool
+	eventRecorder      events.Recorder
+}
+
+type Runnable interface {
+	Run(ctx context.Context, workers int)
+}
+
+const (
+	// Minimal interval between controller resyncs. The controller will detect
+	// new share types in Manila and create StorageClasses for them at least
+	// once per this interval.
+	resyncInterval = 1 * time.Minute
+
+	operatorConditionPrefix = "ManilaController"
+)
+
+func NewController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	informers v1helpers.KubeInformersForNamespaces,
+	openStackClient *openStackClient,
+	csiControllers []Runnable,
+	eventRecorder events.Recorder) factory.Controller {
+
+	scInformer := informers.InformersFor("").Storage().V1().StorageClasses()
+	c := &Controller{
+		operatorClient:     operatorClient,
+		kubeClient:         kubeClient,
+		storageClassLister: scInformer.Lister(),
+		openStackClient:    openStackClient,
+		csiControllers:     csiControllers,
+		eventRecorder:      eventRecorder,
+	}
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(operatorClient).ResyncEvery(resyncInterval).WithInformers(
+		operatorClient.Informer(),
+		scInformer.Informer(),
+	).ToController("ManilaController", eventRecorder)
+}
+
+func (c *Controller) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	klog.V(4).Infof("Manila sync started")
+	defer klog.V(4).Infof("Manila sync finished")
+
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorv1.Managed {
+		return nil
+	}
+
+	shareTypes, err := c.openStackClient.GetShareTypes()
+	if err != nil {
+		switch err.(type) {
+		case *gophercloud.ErrEndpointNotFound:
+			// OpenStack does not support manila, report the operator as disabled
+			return c.setDisabled("This OpenStack does not provide Manila service")
+		default:
+			return err
+		}
+	}
+
+	if len(shareTypes) == 0 {
+		klog.V(4).Infof("Manila does not provide any share types")
+		return c.setDisabled("Manila does not provide any share types")
+	}
+	// Manila has some shares: start the actual CSI driver controller sets
+	if !c.controllersRunning {
+		klog.V(4).Infof("Starting CSI driver controllers")
+		for _, ctrl := range c.csiControllers {
+			go ctrl.Run(ctx, 1)
+		}
+		c.controllersRunning = true
```

Comment:
nit: technically if `c.csiControllers` is empty this should be false

Reply:
they can't be empty...